Add shortcuts and status bar widgets to toggle and set 3D factor

Mates1500 opened the pull request:
Resolves #4621 

Summary of changes
- Add shortcut to toggle 3D (Ctrl+3 keybind by default)
- Add shortcut to decrease 3D factor by 5 (Ctrl+- keybind by default)
- Add shortcut to increase 3D factor by 5 (Ctrl++ keybind by default)
- Add 3D status label to the status bar. Clicking the label toggles the 3D as well.
- Add 3D factor QSpinBox next to the status label. SingleStep set to 5.

One thing that I couldn't for the love of god figure out is how to make the QSpinBox narrower, because it's wider than it needs to be imo. Calling `setMaximumWidth()` nor `setFocusedWidth()` seemed to help. Any suggestions there?

Let me know if there's anything not up to par with the coding style or if I perhaps forgot something.

# Screenshots:
## 3D off
![image](https://user-images.githubusercontent.com/732314/58748345-9e1a2900-8477-11e9-90cc-e2ee7f34fc89.png)
## 3D on
![image](https://user-images.githubusercontent.com/732314/58748348-a7a39100-8477-11e9-90eb-e9ee010b7218.png)


Now that I think about it, maybe we should move these widgets either to the right, or set the other labels to fixed size? Since the speed, framerate and frametime changing all the time causes all the widgets to the left of them to also move due to anchoring.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/citra-emu/citra/4787)
<!-- Reviewable:end -->

zhaowenlan1779 commented on src/citra_qt/configuration/config.cpp:
```diff
@@ -54,21 +54,24 @@ const std::array<std::array<int, 5>, Settings::NativeAnalog::NumAnalogs> Config:
 // QKeySequnce(...).toString() is NOT ALLOWED HERE.
 // This must be in alphabetical order according to action name as it must have the same order as
 // UISetting::values.shortcuts, which is alphabetically ordered.
```
I think you should check this kind tip and change the shortcut entries in the other file as well.

Mates1500 replied:
are you sure about this? those seem to be automatically initialized at line 357
https://github.com/citra-emu/citra/blob/186ffc235f744dad315a603a98cce4597ef0f65f/src/citra_qt/configuration/config.cpp#L350-L360

zhaowenlan1779 replied:
Ah, I see. sorry.

zhaowenlan1779 commented on src/citra_qt/main.cpp:
```diff
@@ -612,6 +670,16 @@ void GMainWindow::ConnectMenuEvents() {
+void GMainWindow::ConnectStatusBarEvents() {
+    connect(ui.action_Toggle_3D, &QAction::triggered, this, &GMainWindow::Toggle3D);
+    connect(status_3d_label, &ClickableLabel::clicked, ui.action_Toggle_3D, &QAction::trigger);
+    connect(factor_3d_spinbox, static_cast<void (QSpinBox::*)(int)>(&QSpinBox::valueChanged),
```
use the qOverload helper.

Mates1500 replied:
fixed

zhaowenlan1779 commented on src/citra_qt/main.cpp:
```diff
@@ -612,6 +670,16 @@ void GMainWindow::ConnectMenuEvents() {
+    connect(ui.action_Toggle_3D, &QAction::triggered, this, &GMainWindow::Toggle3D);
+    connect(status_3d_label, &ClickableLabel::clicked, ui.action_Toggle_3D, &QAction::trigger);
+    connect(factor_3d_spinbox, static_cast<void (QSpinBox::*)(int)>(&QSpinBox::valueChanged),
+            this, [&](int value) {
```
try not to capture everything, just capturing 'this' should be enough.

wwylele replied:
Just to add to this, the reason is that people using `[&]` without care has caused many bugs in the past, like capturing local var by reference and using it out side. We figured that explicitly listing capture can help reducing the the mistake.

Mates1500 replied:
got it, thanks for the heads up, I'll study the C++ lambda captures a little bit more, because I'm not very familiar with them coming from C#
fixed

zhaowenlan1779 commented on src/citra_qt/main.ui:
```diff
@@ -474,6 +474,24 @@
+   <property name="text">
+    <string>Decrease 3D</string>
+   </property>
+  </action>
```
is there a specific reason these have to be actions, rather than just functions, if you are not going to put them in a menu?

plus, the 'increase' and 'decrease' actions don't seem to be connected anywhere...

Mates1500 replied:
completely forgot about these, thanks for the heads up
I've removed all of them and corrected/removed references where needed

Mates1500 commented:
> Please attach a screenshot for UI changes. Thanks!

Got it, I've updated the original post

chris062689 commented:
I do agree we need some kind of a visual toggle. Users already accidentally mash their keyboard and enable all sorts of options and don't know how to turn then off. 

I don't think every option should have a keybind by default. 

Users already accidentally hit the frame advance key, for example, and think their game is frozen.

Mates1500 commented:
> Can the spinbox be replaced with a slider?
> 
> Also, I don't really like the idea of cluttering up the status bar. But I don't have an alternative in mind either.

Here's what a Slider looks like

![image](https://user-images.githubusercontent.com/732314/58758800-1f79c600-8521-11e9-9fb1-a11210ccc5f6.png)
I personally think it does look nicer, but feel the `QSpinBox` is more practical due to being able to enter a specific numeric value. Perhaps a vote could work?

legoj15 commented:
> I personally think it does look nicer, but feel the `QSpinBox` is more practical due to being able to enter a specific numeric value.

I don't see why any standard user would want a specific value.  If it's anyway like how I (and many other people) use the 3DS, I just either have the 3D slider all the way up, or all the way down, never an exact position in-between. I think a slider is perfect; it mirrors the functionality of the 3DS.

Mates1500 commented:
Ok, I've changed the PR implementation to the QSlider

I have also split the space between the widgets even, spread across dynamically dependent on the width of the window. The primary reasoning for that is so that the 3D slider/status text etc don't get moved when the numbers get too high/low, causing the labels to shrink/expand, moving along all the content to the left of it.

# Current screenshots off/on

![image](https://user-images.githubusercontent.com/732314/58908460-ed2ebb00-8710-11e9-8478-f94576004025.png)
![image](https://user-images.githubusercontent.com/732314/58908473-f750b980-8710-11e9-8209-c9293ec10cf6.png)

Currently the slider width scales dynamically with the width of the window. Perhaps some sort of maximum width limit should be set?

B3n30 commented:
please resolve conflicts

adityaruplaha commented:
Instead of splitting them evenly like that, I think moving all toggles/sliders to the right and the messages to the left is a better idea. It presents a clear visual distinction.

One more idea. Can we have a vertical slider popping up (preferably temporarily) right above the toggle? That way we're using less of the available area on the status bar.

If I get time I'll try to construct a screenshot of what I wanna say.

FearlessTobi commented:
Superseded by https://github.com/citra-emu/citra/pull/6277.
